_posts/2016-11-02-improving-page-speed-with-graphql.markdown

```diff
@@ -0,0 +1,113 @@
+
+## Speed Issues
+
+In our publishing CMS called [Positron](http://github.com/artsy/positron), we serve a separate API, database, and front-end from the rest of the Artsy stack. We're also responsible for the delivery and presentation of the content itself. Over the past year we've focused a lot on how the content appears, and now we've opened up time to focus on speed since in the land of content, every second counts.
+
+We split the speed problem in two: `Server Response Time` (server-side) and `Document Interactive Time` (client-side). We currently use Google Analytics to track these [metrics](https://support.google.com/analytics/answer/2383341?hl=en). There are some inaccuracies with GA such as small sampling, geographic noise (countries with different download speeds), and timeouts being counted as zero, but for our initial testing this will suffice for relative measuring.
+
+## Enter GraphQL and JoiQL
+
+Our very own [Craig Spaeth](https://twitter.com/craigspaeth) recently started working on a project called [Mural](https://github.com/muraljs/mural). It's a framework for React and GraphQL. One library that came out of this project is called [JoiQL](http://github.com/muraljs/joiql). It's a The main purpose of JoiQL is to convert [Joi](http://github.com/hapijs/joi) schemas into GraphQL schemas, and vice versa.
+
+We already use Joi in Positron so creating a GraphQL-based endpoint was trivial with JoiQL. Note that while JoiQL is currently used in production, it's still a beta project! Below is an example of how the JoiQL setup works.
+
+
+```javascript
+const joiql = require('../')
+const { object, string, number, array, date } = require('joi')
+const app = require('express')()
+const graphqlHTTP = require('express-graphql')
+
+# Given a Joi schema:
+const Article = object({
+  title: string(),
+  body: string(),
+  published_at: date()
+}).meta({
+  args: { id: number().required() }
+})
+
+# Define api with JoiQL like this:
+const api = joiql({
+  query: {
+    article: Article
+  }
+})
+
+# Resolve the request using a Koa 2 style middleware pattern:
+api.use((ctx, next) => {
+  return new Promise (resolve, reject) ->
+    # Method that fetches an article based on the query
+    findArticle(ctx.req.query.article.args), (err, results) ->
+      ctx.res.article = results.article
+      next()
+      resolve()
+})
+
+# Finally, mount our schema to express:
+app.use('/graphql', graphqlHTTP({
+  schema: api.schema,
+  graphiql: true
+}))
+app.listen(3000, () => console.log('listening on 3000'))
+
+```
+
+You can see how simple it becomes to convert apps that already use Joi with JoiQL. It's less JSON blobs to handle and we also get a nice Koa 2 style middleware pattern that lets us hook into a downstream/upstream flow. This could be useful later on when if we were to get really granular with measuring speed and decide to track the speed of a request manually. We could do something like this:
```

Comment:
@orta I'll take a stab at a slight revision on that language, let me know if this clears things up (also Kana, lmk if this gets across what you intended).

You can see how simple it becomes to convert apps that already use Joi with JoiQL. ~~It's less JSON blobs~~ Joi provides a nice API for building GraphQL schemas with minimal boilerplate and we also get a nice [Koa 2](http://koajs.com/) style middleware pattern that lets us hook into a downstream/upstream flow. This **downstream/upstream flow** could be useful later on when if we were to get really granular with measuring speed and decide to track the speed of a request manually. We could do something like this: